src/main/java/org/truffleruby/core/string/StringNodes.java

```diff
@@ -3374,31 +3423,74 @@ protected Object capitalizeMultiByteComplex(RubyString string, int caseMappingOp
                 @Cached @Shared("dummyEncodingProfile") BranchProfile dummyEncodingProfile,
                 @Cached @Shared("emptyStringProfile") ConditionProfile emptyStringProfile,
                 @Cached @Shared("modifiedProfile") ConditionProfile modifiedProfile) {
-            final Rope rope = string.rope;
-            final Encoding enc = rope.getEncoding();
+            var tstring = string.tstring;
+            var encoding = string.encoding;
 
-            if (enc.isDummy()) {
+            if (encoding.jcoding.isDummy()) {
                 dummyEncodingProfile.enter();
                 throw new RaiseException(
                         getContext(),
-                        coreExceptions().encodingCompatibilityErrorIncompatibleWithOperation(enc, this));
+                        coreExceptions().encodingCompatibilityErrorIncompatibleWithOperation(encoding.jcoding, this));
             }
 
-            if (emptyStringProfile.profile(rope.isEmpty())) {
+            if (emptyStringProfile.profile(tstring.isEmpty())) {
                 return nil;
             }
 
-            final RopeBuilder builder = RopeBuilder.createRopeBuilder(bytesNode.execute(rope), string.encoding);
+            var byteArray = byteArrayNode.execute(tstring, encoding.tencoding);
+
+            // TODO (nirvdrum 26-May-22): Make the byte array builder copy-on-write so we don't eagerly clone the source byte array.
+            var builder = ByteArrayBuilder.create(byteArray);
+
+            var cr = getCodeRange(tstring, encoding.tencoding);
             final boolean modified = StringSupport
-                    .capitalizeMultiByteComplex(enc, codeRangeNode.execute(rope), builder, caseMappingOptions, this);
+                    .capitalizeMultiByteComplex(encoding.jcoding, cr, builder, caseMappingOptions, this);
+
             if (modifiedProfile.profile(modified)) {
-                string.setTString(fromByteArrayNode.execute(builder.getBytes(), string.encoding.tencoding));
+                string.setTString(makeTString(builder.getUnsafeBytes(), encoding.tencoding));
                 return string;
             } else {
                 return nil;
             }
         }
 
+        private byte[] copyByteArray(AbstractTruffleString string, TruffleString.Encoding encoding) {
```

Comment:
This looks like a good candidate to move to TStringUtils to make CopyToByteArrayNode a little easier to use.

Maybe we should have a `byte[] execute(AbstractTruffleString a, Encoding expectedEncoding)` in CopyToByteArrayNode directly, cc @djoooooe, that'd be quite convenient.

Reply:
Done in fc196fcc0fcc94dfceed091a865fe09de40c7f52. I didn't see a need for an uncached version yet, but we can easily add that if the need arises.